### DISCUSSION TITLE
Soundfond bank/patch remap within FluidSynth

### DISCUSSION THREAD

**Discussion by trolley813** [FluidSynth/fluidsynth#809] [Ideas]
The main idea is to allow arbitrarily remap presets (using e.g. a map config file), especially when using multiple soundfonts at once

Why it may be needed:
For example, GS (mostly extended versions like SC-88) and XG have their presets largely intersecting (outside of GM, namely, presets like "12-String Guitar", "Jazz Bass", "Rotary Organ" etc.), but unfortunately, they are available at different bank numbers. Since there exist some free soundfonts which allow complete (or nearly complete) coverage of these standards (for example, [this one](https://musical-artifacts.com/artifacts/909)), it would be comfortable to use single set of soundfonts to play MIDI files of both types (without modifying the SF2s themselves). Also,  it would be a more convenient way to manipulate banks (rather than already-present bank offset feature, which may not work for standard MIDI files since it only allows offsetting all the banks present in an SF2 file by a fixed amount).
Another potential use-case is to play non-GM-compliant MIDI files with GM soundfonts and vice versa (by mapping not only banks, but also program numbers).

How it can work:
For example, a GS soundfont (loaded into FluidSynth in slot (SFID) 0) may contain the "Mute Picked Bass" preset on bank 8, program number 35 (using 1-128 range).
The map file could contain the following settings (in some format):
```
Map the program #35 in bank #1024 (which corresponds to MSB bank 8 when the bank select mode is set to MMA) to SFID 0, program #35, bank #8
Map the program #35 in bank #28 (bank number for the matching preset "MutePkBa" in XG) to SFID 0, program #35, bank #8
```
So, when the bank select mode is set to MMA, both GS and XG files containing that preset could play correctly. One should be able to even mix GS and XG patches in one file.

How it can be implemented:
Actually, it's very easily to implement - one just needs to find the preset in the map (if remapping is enabled) when a program change event arrives, similarly to the already implemented bank offset, with reasonable defaults, such as setting to bank 0 when the requested preset is not found in the map (of course, one will need additionally the code to parse the map file). Since the feature is designed to be completely optional, it should not introduce any backwards incompatibility (either API or ABI).

**Comment by mawe42** [comment]
It probably wouldn't be hard to implement, I agree. The question I'm asking myself is: how many people would actually use this feature? From your description I'm not quite clear if that is just an idea you had and you think other people might need it, or if it is a solution to a problem that you have personally.

If we were to implement this, then I would vote against a separate map file and rather add shell commands to add or change mappings. And in my opinion, the feature would need to be available via the API as well.

Cheers
Marcus

**Reply by yukulele** [reply]
I also nead it!

my use case:

I use `fluidR3_GM.sf2` as main soundfont. Most of the instruments suit me except some.

I want for example to use a `recorder.sf2` to replace the fluidR3 recorder:

```shell
> load recorder.sf2
loaded SoundFont has ID 2
> inst 2
000-000 Recorder
000-001 Recorder N1
000-002 Recorder N2
```
Unfortunately recorder sound replace patch 0, 1 & 2 and piano now sound like a recorder...

a `remap` command should help.

spec suggestion:

```shell
// remap inst patch dest
> remap 2 000-001 none
> remap 2 000-002 000-074
> remap 2 000-003 none
```
or

```shell
// remap inst patch dest
> remap 2 000 none
> remap 2 000-002 000-074
```

**Reply by trolley813** [reply]
@mawe42 yes, it is mostly an idea, so that's the reason I posted it in _Ideas_. As @yukulele said, there are non-GM-compatible soundfonts which cannot be used with GM ones using bank offset only.
And I agree to both adding the shell commands as a primary method (loading from a file could be an auxiliary way for convenience, when there are hundreds of mappings) and exposing this in API.

**Reply by yukulele** [reply]
What do you think  @mawe42?

**Comment by jimhen3ry** [comment]
If I remember correctly, if you load multiple soundfonts into the same bank, later loaded instruments replace earlier loaded instruments. That is what I understand yukulele to say happens with the example of the recorder.sf2 patches replacing piano patches. 

Would making a modified version of a soundfont being used to replace instruments in an earlier loaded soundfont to renumber the presets to merge properly be an acceptable answer? That is easy to do with a program like Polyphone. Or is there a need to be able to change the preset numbers at the time FluidSynth is being set up?

**Reply by yukulele** [reply]
> ...later loaded instruments replace earlier loaded instruments...

Indeed, that's how it works.

> Would making a modified version of a soundfont being used to replace instruments in an earlier loaded soundfont to renumber the presets to merge properly be an acceptable answer?...

This is not a really satisfying solution for me because it forces to manage a modified version of my sf2 for each configuration, moreover the use of Polyphone is a rather cumbersome process compared to a command line tool.
